File: hot/leetcode/78.子集/78.子集.go
package main

import "fmt"

func subsets(nums []int) [][]int {
	n := len(nums)
	// 路径列表
	track := make([]int, 0, n)
	res := [][]int{}

	var backTrack func(start int)
	backTrack = func(start int) {
		// 把路径添加到结果集，生成新的数组的方式
		res = append(res, append(make([]int, 0, n), track...))
		// 遍历选择
		for i := start; i < n; i++ {
			// 做选择
			track = append(track, nums[i])
			// 递归
			backTrack(i + 1)
			// 取消选择
			track = track[:len(track)-1]
		}
	}
	backTrack(0)
	return res
}

func main() {
	nums := []int{1, 2, 3}
	result := subsets(nums)
	fmt.Println(result)
}

File: hot/leetcode/78.子集/78.子集.py
from typing import List


class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        # 路径
        track = []
        res = []
        n = len(nums)

        def back_track(start):
            # 满足结束条件
            res.append(list(track))
            # 遍历 选择
            for i in range(start, n):
                # 做选择
                track.append(nums[i])
                # 递归
                back_track(i + 1)
                # 撤销选择
                track.pop()

        back_track(0)
        return res


if __name__ == '__main__':
    nums = [1, 2, 3]
    result = Solution().subsets(nums)
    print(result)
